File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys 
input = sys.stdin.readline

N,M,R = map(int,input().split())
G=[[] for _ in range(N + 1)]

for i in range(M):
    u,v = map(int,input().split())
    G[u].append(v)
    G[v].append(u)

for i in G : i.sort()

visited = [0] * (N + 1)
visited[R] = 1
D = [0] * (N + 1)
D[R] = 1
Q = [R]
order = 2

while Q:
    V = Q.pop(0)
    for w in G[V]:
        if visited[w] == 0:
            visited[w] = 1
            Q.append(w)
            D[w] = order
            order += 1

for i in range(1,N+1):
    print(D[i])

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const fs = require("fs")
let input = fs.readFileSync(filePath).toString().trim().split("\n")

class Queue{
  head = 0
  tail = 0
  list = []
  constructor(){}

  append(value){
    this.list[this.tail++] = value
  }
  popleft(){
    if(this.isEmpty()){
      return false
    }
    return this.list[this.head++]
  }
  isEmpty(){
    return this.head === this.tail ? true : false
  }
}

const [n, m, r] = input[0].split(" ").map((e) => Number(e))
const G = Array.from({length: n + 1}, () => [])
const vi = new Array(n + 1).fill(0)
let order = 1

for (let i = 1; i <= m; i++){
  const [sp,ep] = input[i].split(" ").map(Number)
  G[sp].push(ep)
  G[ep].push(sp)
}

G.forEach((line) => line.sort((a,b) => a - b))

let Q = new Queue()
Q.append(r)
vi[r] = 1

while(!Q.isEmpty()) {
  const sp = Q.popleft()
  for (const np of G[sp]){
    if(!vi[np]) {
      Q.append(np)
      vi[np] = ++order
    }
  }
}

console.log(vi.slice(1).join("\n"))